hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockInputStream.java

```diff
@@ -187,19 +192,19 @@ public synchronized void initialize() throws IOException {
     }
   }
 
+  /**
+   * Check if this exception is because datanodes are not reachable.
+   */
+  private boolean isConnectivityIssue(IOException ex) {
+    return Status.fromThrowable(ex).getCode() == Status.UNAVAILABLE.getCode();
+  }
+
   private void refreshPipeline(IOException cause) throws IOException {
     LOG.info("Unable to read information for block {} from pipeline {}: {}",
         blockID, pipeline.getId(), cause.getMessage());
     if (refreshPipelineFunction != null) {
       LOG.debug("Re-fetching pipeline for block {}", blockID);
-      Pipeline newPipeline = refreshPipelineFunction.apply(blockID);
-      if (newPipeline == null || newPipeline.sameDatanodes(pipeline)) {
-        LOG.warn("No new pipeline for block {}", blockID);
-        throw cause;
-      } else {
-        LOG.debug("New pipeline got for block {}", blockID);
-        this.pipeline = newPipeline;
-      }
```

Comment:
@duongkame May I ask why the check for null/same pipeline was removed?

Reply:
@adoroszlai, sorry I missed this. My intention was to remove the check for the same pipeline because it just doesn't make sense and fails some test cases. (Client talks to datanode and get an error, it then recontacts to OM and SCM and SCM say the current pipeline is ok, hence, the previous error with datanode may be just a temporary one, so just retry if the policy still allows).

I wasn't aware of any case where the returned pipeline can be null, think I talked to someone to confirm it as well. Obviously you've caught a problem with it, as per HDDS-7982. Thanks.

Reply:
No problem.  The fix for HDDS-7982 restores the null check, but avoids throwing exception for same pipeline case.

On the positive side, this change has also fixed HDDS-7089 where read after datanode restart was failing intermittently.  It turned out that `sameDatanodes` ignores IP address change, only checks UUIDs.